backtracking/wildcard_matching.cpp

```diff
@@ -0,0 +1,143 @@
+ * @details
+ * Problem Statement: 
+ * Given an matching string and a pattern, implement wildcard pattern matching with support for '?' and '*'. '?' Matches any single character. '*' Matches any sequence of characters (including the empty sequence). The matching should cover the entire matching string (not partial). Task is to determine if the pattern matches with the matching string
+ * @author [Swastika Gupta](https://github.com/Swastyy)
+ */
+
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+#include <vector>    /// for std::vector
+
+/**
+ * @namespace backtracking
+ * @brief Backtracking algorithms
+ */
+namespace backtracking {
+/**
+ * @namespace wildcard_matching
+ * @brief Functions for the [Wildcard Matching](https://www.geeksforgeeks.org/wildcard-pattern-matching/) problem.
+ */
+namespace wildcard_matching {
+/**
+ * @brief The main function implements if pattern can be matched with given string
+ * @param s is the given matching string
+ * @param p is the given pattern
+ * @param pos1 is the starting index
+ * @param pos2 is the last index
+ * @returns 1 if pattern matches with matching string otherwise 0
+ */
+std::vector<std::vector<int>>dpTable(1000, std::vector<int>(1000, -1));
+int32_t wildcard_matching(std::string s, std::string p, int pos1, int pos2)
+{       
+        int n = s.length();
+        int m = p.length();
+	    // matching is successfull if both strings are done
+        if(pos1 == n && pos2 == m)
+		    {
+            return 1;
+		    }
+
+		// matching is unsuccessfull if pattern is not finished but matching string is
+        if(pos1 != n && pos2 == m)
+		    {
+            return 0;
+		    }
+
+		// all the remaining characters of patterns must be * inorder to match with finished string
+        if(pos1 == n && pos2 != m)
+        {
+            while(pos2 < m && p[pos2] == '*')
+			      {
+                pos2++;
+			      }
+
+            return pos2 == m;
+        }
+
+		// if already calculted for these positions
+        if(dpTable[pos1][pos2] != -1)
+		    {
+            return dpTable[pos1][pos2];
+		    }
+
+		// if the characters are same just go ahead in both the string
+        if(s[pos1] == p[pos2])
+		    {
+            return dpTable[pos1][pos2] = wildcard_matching(s, p, pos1 + 1, pos2 + 1);
+		    }
+
+        else
+        {
+		// can only single character
+            if(p[pos2] == '?')
+			      {
+                return dpTable[pos1][pos2] = wildcard_matching(s, p, pos1 + 1, pos2 + 1);
+			      }
+			// have choice either to match one or more charcters
+            else if(p[pos2] == '*')
+			      {
+                return dpTable[pos1][pos2] = wildcard_matching(s, p, pos1, pos2 + 1) || wildcard_matching(s, p, pos1 + 1, pos2);
+			      }
+			// not possible to match
+            else
+			      {
+                return dpTable[pos1][pos2] = 0;
+			      }
+        }
```

Comment:
Some of the indentations here are looking wrong. Please fix that issue.

Reply:
But they are correctly indented when I open it. 😕